kubernetes_state/check.py

```diff
@@ -326,6 +376,22 @@ def kube_node_spec_unschedulable(self, message, **kwargs):
         else:
             self.log.error("Metric type %s unsupported for metric %s" % (message.type, message.name))
 
+    def kube_pod_container_status_waiting_reason(self, message, **kwargs):
```

Comment:
For these two metrics, I'm not sure whether we should tag by podname, or just increment a node counter. @hkaj what's your opinion on that?

Reply:
We need the pod name, node name and reason for this metric. One reason a pod can be waiting is `ErrImagePull`, as a user I'd like to know if I fat-fingered the image name in my pod spec and all the pods of a same family are failing because of that.

I would also like to know if a node is making all its containers wait because it can't pull images or its containers are stuck in creation.

Now how we expose this info is up to discussion. I think the metric name should be `self.NAMESPACE + '.pod_container_status_waiting.' + reason`. Reasons are finite and hardcoded here: https://github.com/kubernetes/kube-state-metrics/blob/b26cfb26fe5e858f8033a641a321abef25496c62/collectors/pod.go#L38 so it should be fine. If we add pods and nodes as tags we should be able to group/filter across the cluster quite easily. I'm open to other options though @CharlyF & @xvello

Reply:
@hkaj I liked the option of not adding reason in the metric name so that one could monitor the metric reason and be alerted if there are more than X pods waiting because of {{reason}} on {{node.name}}.
Less cumbersome to set for a user imo.

Reply:
Yeah that's a good point, then what do you think of removing `.reason` from the metric name? `pod.containers_waiting` with the `reason` tag?

Reply:
So I think `reason` should remain in the name to distinguish the metric from https://github.com/DataDog/integrations-core/blob/da26a02d67a5fb3639b44abc319a64d358b80ae5/kubernetes_state/check.py#L74 (so that both can be used)

Reply:
hmm shouldn't we just add the reason tag to the existing metric then? Does the new one add any value?

Reply:
I actually get the point - Indeed the .reason will be at one, on the tag representing the  reason whereas the regular is at 1 if the container is waiting.
https://github.com/kubernetes/kube-state-metrics/blob/master/collectors/pod_test.go#L284-L292.

Now, I can't just remove `kube_pod_container_status_waiting` and I don't know if using `kube_pod_container_status_waiting_reason` to override or even replace `container.waiting` is a good idea because this metric became available with 1.1.0 so people monitoring the metric already might start missing it. Thoughts ?

Reply:
could we keep the existing metric, but instead of adding the new one we match its pod with the existing metric and just add the reason as a tag to it?

Reply:
Best solution I came up with is adding a meta crawl first https://github.com/DataDog/dd-agent/pull/3583

Reply:
Why did I implement it this way:
The prometheus process_metric method goes through all the items in ignored, metrics_mapper and then getattr (i.e. methods in k8s-state check with the metric name as a method name).
While processing the waiting_reason metric, we don't have access to the message of waiting (i.e. pod name). 
The approach is: If we see the new waiting_reason, populate the old metric name and add the reason as a tag.
We can't use a boolean because we can't trust the order in which the metrics are coming from prometheus and hence, might miss the _reason and compute the old one before.
*The solution:*
We can't use a boolean for all the new metrics because the waiting_reason might be exposed but terminated_reason might not and we don't want to miss terminated.
Secondly, from my testings, the decision of whether to compute the new/old metrics or not *has* to be before processing the metrics in the kube-state check because getattr is non blocking and metrics are missed.
Using a dict of meta metrics will make this re-usable should we want to process several times the prometheus output to associate more metadata with the underlying pods